### PR TITLE
Resolving Code Review Comments

### PR DESCRIPTION
**Pull request by moniercoding**
Reverted to sending payload and header separately while maintaining the use of serialized attributes.

### CODE REVIEW

File: lib/include/up-client-zenoh-cpp/session/zenohSessionManager.h [hunk 1]
```diff
@@ -28,8 +28,8 @@
 #include <atomic>
 #include <zenoh.h>
 #include <up-cpp/transport/datamodel/UPayload.h>
-#include <up-cpp/transport/datamodel/UAttributes.h>
 #include <up-core-api/ustatus.pb.h>
+#include <up-core-api/uattributes.pb.h>
```
Comment (MelamudMichael) [inline lib/include/up-client-zenoh-cpp/session/zenohSessionManager.h:32]: I think that you can remove the include

File: lib/src/zenohRpcClient.cpp [hunk 2]
```diff
@@ -130,48 +126,43 @@ std::future<UPayload> ZenohRpcClient::invokeMethod(const UUri &uri,
...
     }
 
-    z_bytes_t bytes;
+    z_owned_bytes_map_t map = z_bytes_map_new();
+    z_bytes_t attrBytes = {.len = serializedAttributes.size(), .start = serializedAttributes.data()};
+    z_bytes_map_insert_by_alias(&map, z_bytes_new("attributes"), attrBytes);
 
-    bytes.len = header.size();
-    bytes.start = reinterpret_cast<const uint8_t*>(header.data());
+    z_bytes_t payloadBytes = {.len = payload.size(), .start = reinterpret_cast<uint8_t*>(payload.data())};
```
Comment (MelamudMichael) [inline lib/src/zenohRpcClient.cpp:141]: why do you put the payload in the header? it should be sent as a payload ...

File: lib/src/zenohUTransport.cpp [hunk 3]
```diff
@@ -178,78 +178,54 @@ UStatus ZenohUTransport::send(const UUri &uri,
...
-        if (attributes.serializationHint().has_value()) {
-            if (UCode::OK != mapEncoding(attributes.serializationHint().value(), options.encoding)) {
-                spdlog::error("mapEncoding failure");
-                z_drop(z_move(map));
-                break;
-            }
-        } else {
-            options.encoding = z_encoding(Z_ENCODING_PREFIX_TEXT_PLAIN, nullptr);
-        }
+        options.encoding = z_encoding(Z_ENCODING_PREFIX_APP_PROTOBUF, nullptr);
```
Comment (MelamudMichael) [inline lib/src/zenohUTransport.cpp:228]: not needed , the payload is not encoded as protobuf

File: lib/src/zenohUTransport.cpp [hunk 4]
```diff
@@ -178,78 +178,54 @@ UStatus ZenohUTransport::send(const UUri &uri,
...
-            spdlog::error("Wrong message type = {}", UMessageTypeToString(attributes.type()).value());
+            spdlog::error("Wrong message type = {}", static_cast<int>(attributes.type()));
             return UCode::INVALID_ARGUMENT;
         }
 
-        /* get hash and check if the publisher for the URI is already exists */
-        auto uriHash = std::hash<std::string>{}(LongUriSerializer::serialize(uri)); 
-        auto handleInfo = pubHandleMap_.find(uriHash); 
+        // Serializing UAttributes
+        size_t attrSize = attributes.ByteSizeLong();
```
Comment (MelamudMichael) [inline lib/src/zenohUTransport.cpp:193]: move this code to be just before the actual publishing

File: lib/src/zenohUTransport.cpp [hunk 5]
```diff
@@ -293,44 +266,33 @@ UCode ZenohUTransport::sendQueryable(const UUri &uri,
...
-    if (0 != z_query_reply(&lquery, z_query_keyexpr(&lquery), payload.data(), payload.size(), &options)) {
+    if (0 != z_query_reply(z_loan(&query), z_query_keyexpr(&query), payload.data(), payload.size(), &options)) {
         spdlog::error("z_query_reply failed");
         z_drop(z_move(map));
         return UCode::INTERNAL;
     }
 
-    auto keyStr = z_keyexpr_to_string(z_query_keyexpr(&lquery));
-
-    z_drop(z_move(keyStr));
```
Comment (MelamudMichael) [inline lib/src/zenohUTransport.cpp:325]: why did you delete this code?

File: lib/src/zenohUTransport.cpp [hunk 6]
```diff
@@ -22,14 +22,14 @@
  * SPDX-License-Identifier: Apache-2.0
```
Comment (MelamudMichael) [inline lib/src/zenohUTransport.cpp:22]: Did you run the sample code to see that everything is working?

**Review by MelamudMichael — CHANGES_REQUESTED** [review]

**Comment by PLeVasseur** [comment]
@moniercoding -- hey there :slightly_smiling_face: 

Seemed like this PR overlapped with #16 and #16 looked more recent. Should this one be closed in favor of that one or are different things intended for each?

**Comment by moniercoding** [comment]
> @moniercoding -- hey there 🙂
> 
> Seemed like this PR overlapped with #16 and #16 looked more recent. Should this one be closed in favor of that one or are different things intended for each?

Hey @PLeVasseur, yes This should be closed in favor of #16.

**Pull request closed without merging**